### PR TITLE
Additional RPC scaffolding

### PR DESCRIPTION
**Pull request by aspect**
This PR includes message structs and routing for additional RPC functions.

### CODE REVIEW

**Review by tiram88 — CHANGES_REQUESTED** [review]
- I addressed you questions.
- In substance, almost all good.
- Nice macros. The paste crate is of great help, I'm fan!
- Some more cleanup is welcome.
- Formally just a few points to better comply with our coding (though non explicit) rules, if you don't mind.

Note that the mining demo is broken but this does not come from your code. I checked in my mempool branch and got same result. The problem lies in the consensus (consensus\src\pipeline\virtual_processor\processor.rs).

File: consensus/core/src/networktype.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
+use addresses::Prefix;
```
[the remark this answers is not in the document]
Comment (aspect) [reply consensus/core/src/networktype.rs:1]: I am following rust standards (that recommend against this) followed in say consensus module, such as blockhash and blockstatus.

File: rpc/core/src/api/rpc.rs [hunk 2]
```diff
@@ -44,132 +58,134 @@ pub trait RpcApi: Sync + Send {
...
+        self.get_selected_tip_hash_call(GetSelectedTipHashRequest {}).await
+    }
+    async fn get_selected_tip_hash_call(&self, request: GetSelectedTipHashRequest) -> RpcResult<GetSelectedTipHashResponse>;
 
-    // async fn get_mempool_entry(
-    //     &self,
-    //     request:  GetMempoolEntryRequest
-    // ) -> RpcResult<GetMempoolEntriesResponse>;
+    ///
+    async fn get_mempool_entry_call(&self, request: GetMempoolEntryRequest) -> RpcResult<GetMempoolEntryResponse>;
```
[the remark this answers is not in the document]
Comment (aspect) [reply rpc/core/src/api/rpc.rs:74]: I was just not sure if this is something that should be done.  In my original design this is not something I had.  Do you intend on having each function accept parametric arguments?  and also have message structs have constructors?  that's a lot of redundant code, whereas I am against constructors and RPC functions arguments (for all but fns that have no args), so I am somewhat staying away from all that overhead :)

File: rpc/core/src/error.rs [hunk 3]
```diff
@@ -48,5 +51,3 @@ impl From<&str> for RpcError {
         RpcError::General(value.to_string())
     }
 }
-
-pub type RpcResult<T> = std::result::Result<T, crate::RpcError>;
```
[the remark this answers is not in the document]
Comment (aspect) [reply rpc/core/src/error.rs:52]: Moved the `Result` back to `error`.  In my practice, there are various benefits of segregating results from errors and many mainstream crates do this. this is more of an "unformulated rust practice", but varies from crate to crate.

File: rpc/core/src/model/networktype.rs [hunk 4]
```diff
@@ -0,0 +1 @@
+pub type RpcNetworkType = consensus_core::networktype::NetworkType;
```
[the remark this answers is not in the document]
Comment (aspect) [reply rpc/core/src/model/networktype.rs:1]: I followed naming convention in consensus such as `blockhash.rs` etc.  Moved this discussion to TG for the time being

File: rpc/core/src/model/peer.rs [hunk 5]
```diff
@@ -0,0 +1,27 @@
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, Serialize, Deserialize, BorshSerialize, BorshDeserialize, BorshSchema)]
+pub struct RpcPeerAddress {
+    // FIXME type
```
[the remark this answers is not in the document]
Comment (aspect) [reply rpc/core/src/model/peer.rs:6]: Oh, I seem to have forgotten to bring that up in telegram - I left FIXME markers specifically for you to fix types where I don't know what the types should be.

File: rpc/grpc/src/client/route.rs [hunk 6]
```diff
@@ -0,0 +1,34 @@
+macro_rules! route {
```
[the remark this answers is not in the document]
Comment (aspect) [reply rpc/grpc/src/client/route.rs:1]: Note: this macro is an extract of the function generated by the `#[async_trait]` as implementing a simple fn definition using basic macros was not working combined with the `#[async_trait]` attribute macro.

**Review by michaelsutton — APPROVED** [review]

**Pull request merged**